File: app/src/main/java/com/mytlogos/enterprise/background/api/MediumApi.java
package com.mytlogos.enterprise.background.api;

import com.mytlogos.enterprise.background.api.model.ClientMedium;
import com.mytlogos.enterprise.background.api.model.ClientMediumInWait;
import com.mytlogos.enterprise.background.api.model.ClientSimpleMedium;

import java.util.List;
import java.util.Map;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.QueryMap;
import retrofit2.http.Url;

interface MediumApi {

    @GET
    Call<ClientMedium> getMedium(@Url String url, @QueryMap Map<String, Object> body);

    @GET
    Call<List<ClientMedium>> getMedia(@Url String url, @QueryMap Map<String, Object> body);

    @GET("{start}/all")
    Call<List<Integer>> getAllMedia(@Path(value = "start", encoded = true) String url, @QueryMap Map<String, Object> body);

    @GET("{start}/unused")
    Call<List<ClientMediumInWait>> getMediumInWait(@Path(value = "start", encoded = true) String url, @QueryMap Map<String, Object> body);

    @PUT("{start}/unused")
    Call<Boolean> consumeMediumInWait(@Path(value = "start", encoded = true) String url, @Body Map<String, Object> body);

    @POST("{start}/create")
    Call<ClientMedium> createFromMediumInWait(@Path(value = "start", encoded = true) String url, @Body Map<String, Object> body);

    @POST
    Call<ClientSimpleMedium> addMedia(@Url String url, @Body Map<String, Object> body);

    @PUT
    Call<Boolean> updateMedia(@Url String url, @Body Map<String, Object> body);
}

File: app/src/main/java/com/mytlogos/enterprise/background/api/MediumApi.kt
package com.mytlogos.enterprise.background.api

import com.mytlogos.enterprise.background.api.model.ClientMedium
import com.mytlogos.enterprise.background.api.model.ClientMediumInWait
import com.mytlogos.enterprise.background.api.model.ClientSimpleMedium
import retrofit2.Response
import retrofit2.http.*

internal interface MediumApi {
    @GET
    suspend fun getMedium(@Url url: String, @QueryMap body: MutableMap<String, Any?>): Response<ClientMedium>

    @GET
    suspend fun getMedia(@Url url: String, @QueryMap body: MutableMap<String, Any?>): Response<List<ClientMedium>>

    @GET("{start}/all")
    suspend fun  getAllMedia(@Path(value = "start", encoded = true) url: String, @QueryMap body: MutableMap<String, Any?>): Response<List<Int>>

    @GET("{start}/unused")
    suspend fun  getMediumInWait(@Path(value = "start", encoded = true) url: String, @QueryMap body: MutableMap<String, Any?>): Response<List<ClientMediumInWait>>

    @PUT("{start}/unused")
    suspend fun consumeMediumInWait(@Path(value = "start", encoded = true) url: String, @Body body: MutableMap<String, Any?>): Response<Boolean>

    @POST("{start}/create")
    suspend fun createFromMediumInWait(@Path(value = "start", encoded = true) url: String, @Body body: MutableMap<String, Any?>): Response<ClientMedium>

    @POST
    suspend fun  addMedia(@Url url: String, @Body body: MutableMap<String, Any?>): Response<ClientSimpleMedium>

    @PUT
    suspend fun updateMedia(@Url url: String, @Body body: MutableMap<String, Any?>): Response<Boolean>
}